content/articles/email-parser-python/index.md

```diff
@@ -0,0 +1,172 @@
+### Parsing emails in Python
+
+Perhaps you are a machine learning engineer trying to build an email spam classifier. You may want to get some way of pre-processing the emails or might be trying to look for  correlations between random emails. You will, in one way or another, have to parse the emails first. We will look at how to do that using Python. Python has the `email` module containing methods to help us achieve this.
+
+### Prerequisites
+
+The following are the prerequisites for this article:
+- An understanding of Python.
+- Jupyter notebook installed or access to Google Colab.
+- An understanding of Python.
+- Jupyter notebook installed or access to Google Colab.
+
+> We will use Jupyter notebooks in this article. You may use regular Python files for the code executions.
+
+A Colab notebook with the codes is found [here](https://colab.research.google.com/drive/1nDz58G4cDukqukOlPRVST-jKBK0AyMwa?usp=sharing).
+
+### Getting started
+
+We will parse emails from the _SpamAssassin_ website, and then look at their structure and contents. The site contains spam and normal emails. We will try and look at their structures using the metrics highlighted. 
+
+We will import the packages we need in a new cell then add the code for getting the emails from the _SpamAssassin_ website.
+
+```python
+import os
+import tarfile
+import urllib.request
+import email
+import email.policy
+from collections import Counter
+import urlextract
+
+#the root url
+EMAILS_URL_ROOT = "http://spamassassin.apache.org/old/publiccorpus/"
+#the emails url
+NORMALS_URL = EMAILS_URL_ROOT + "20030228_easy_ham.tar.bz2"
+SPAMS_URL = EMAILS_URL_ROOT + "20030228_spam.tar.bz2"
+#datasets path
+SPAM_PATH = os.path.join("datasets", "spam")
+
+#method for fetching the emails from the url
+def fetch_emails(NORMALS_URL=NORMALS_URL, SPAMS_URL=SPAMS_URL, spams_path=SPAM_PATH):
+  #creating a directory
+    if not os.path.isdir(spams_path):
+        os.makedirs(spams_path)
+        for filename, url in (("ham.tar.bz2", NORMALS_URL),("spam.tar.bz2", SPAMS_URL)):
+            path = os.path.join(spams_path, filename)
+            #checking if there is a file
+            if not os.path.isfile(path):
+                urllib.request.urlretrieve(url, path)
+                #extracting
+            email_tar_file = tarfile.open(path)
+            email_tar_file.extractall(path=spams_path)
+            email_tar_file.close()
+```
+We set the paths and the URLs as constants. Then, in the `fetch_emails()` method, we check if the directory is present using the `isdir()` method of the `os` module. If the directory is not present, we create a new one using the `makedirs()` method.
+ 
+We create two paths needed for normal and spam emails to store the emails.
```

Comment:
It is better to call the classification task spam vs. ham instead of spam vs. normal emails. Change the name of folders and all references to ham messages as normal messages.